File: emotion-back/content/src/main/java/com/example/content/mapper/AlbumMapper.java
package com.example.content.mapper;

import com.example.content.dtos.album.AlbumCreateDto;
import com.example.content.dtos.album.AlbumResponseDto;
import com.example.content.dtos.album.AlbumShortDto;
import com.example.content.dtos.album.AlbumUpdateDto;
import com.example.content.entities.Album;
import org.mapstruct.*;

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface AlbumMapper {
    Album toEntity(AlbumCreateDto dto);

    AlbumResponseDto toDto(Album album);

    @Mapping(target = "notesCount",
            expression = "java(album.getNotes() != null ? album.getNotes().size() : 0)")
    AlbumShortDto toShortDto(Album album);

    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "userId", ignore = true)
    @Mapping(target = "notes", ignore = true)
    @Mapping(target = "topics", ignore = true)
    void updateEntityFromDto(AlbumUpdateDto dto, @MappingTarget Album album);
}

File: Android/app/src/main/java/com/example/emotionapp/data/mappers/AlbumMapper.kt
package com.example.emotionapp.data.mappers

import com.example.emotionapp.data.dtos.AlbumResponseDTO
import com.example.emotionapp.data.dtos.AlbumShortDTO
import com.example.emotionapp.domain.entities.Album
import com.example.emotionapp.domain.entities.AlbumSummary

fun AlbumResponseDTO.toDomain(): Album = Album(
    id = id,
    title = title,
    description = description,
    createdAt = createdAt,
    topics = topics.map { it.toDomain() },
    notesCount = null
)

fun Album.toDTO(): AlbumResponseDTO = AlbumResponseDTO(
    id = id,
    title = title,
    description = description,
    createdAt = createdAt,
    topics = topics.map { it.toDTO() },
)

fun AlbumShortDTO.toDomain(): AlbumSummary = AlbumSummary(
    id = id,
    title = title,
    description = description,
    notesCount = notesCount
)
